kinder/cmd/kinder/create/cluster/kubeadmconfig.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/util/version"
+)
+
+// getKubeadmConfigPatches returns the kubeadm config patches for the Kubernetes initVersion
+func getKubeadmConfigPatches(initVersion string) (string, string, string, error) {
+	// gets the kubeadm config version corresponding to a Kubernetes initVersion
+	initConfigVersion, err := getKubeadmConfigVersion(initVersion)
+	if err != nil {
+		return "", "", "", err
+	}
+
+	// select the patches for the kubeadm config version
+	log.Infof("Preparing kubeadm config patches for %s version", initConfigVersion)
+	switch initConfigVersion {
+	case "v1beta2":
+		return kubeDNSPatchv1beta2, calicoPatchv1beta2, externalEtcdPatchv1beta2, nil
+	case "v1beta1":
+		return kubeDNSPatchv1beta1, calicoPatchv1beta1, externalEtcdPatchv1beta1, nil
+	case "v1alpha3":
+		return kubeDNSPatchv1alpha3, calicoPatchv1alpha3, externalEtcdPatchv1alpha3, nil
+	case "v1alpha2":
+		return kubeDNSPatchv1alpha2, calicoPatchv1alpha2, externalEtcdPatchv1alpha2, nil
+	}
+
+	return "", "", "", errors.Errorf("unknown kubeadm config version: %s", initConfigVersion)
+}
+
+// getKubeadmConfigVersion returns the kubeadm config version corresponding to a Kubernetes initVersion
+func getKubeadmConfigVersion(initVersion string) (string, error) {
+	// parses the initVersion
+	vS, err := version.ParseSemantic(initVersion)
+	if err != nil {
+		return "", errors.Wrapf(err, "%s is not a valid version", initVersion)
+	}
+
+	// returns the corresponding kubeadm config version
+	// nb v1alpha1 (that is Kubernetes v1.10.0) is out of support
+
+	if vS.LessThan(version.MustParseSemantic("v1.12.0-0")) {
+		return "v1alpha2", nil
+	}
+
+	if vS.LessThan(version.MustParseSemantic("v1.13.0-0")) {
+		return "v1alpha3", nil
+	}
+
+	if vS.LessThan(version.MustParseSemantic("v1.15.0-0")) {
+		return "v1beta1", nil
+	}
```

Comment:
unfortunately here we are facing the same problem as kubernetes-anywhere, kind, cluster-api and possibly other tools faced with kubeadm. there is no machine-readable mapping between kubeadm version and preferred kubeadm API, so this code has to be maintained manually.

Reply:
That's true, but all the alternatives that I can think of have downsides too...

Reply:
it's just a fundamental problem of downstream consumers of kubeadm.